File: Codeforces/Div 3/Codeforces Round 988/2037E.py
import sys
from functools import cache
sys.setrecursionlimit(int(1e5))

def query(l, r):
    print(f"? {l} {r}", flush=True)
    return int(input())

def answer(ans):
    print(f"! {ans}", flush=True)

t = int(input())
for _ in range(t):
    n = int(input())

    A = [0] * (n + 2)
    f = [0] * (n + 2)
    for i in range(1, n):
        f[i] = query(i, n)
            
    D = [0] * (n + 2) # D[i] 表示以 i 為開頭的子序列數量 
    for i in range(1, n):
        D[i] = f[i] - f[i + 1]

    s = [''] * (n + 1)
    @cache
    def dfs(i, cnt):
        if i == 0:
            return (1, ''.join(s[1:n+1]))
        if D[i] < 0:
            return (0, '')

        if D[i] > 0:
            s[i] = '0'
            if D[i] != cnt:
                return (0, '')
            possible, res = dfs(i - 1, cnt)
            if possible != 1:
                res = ''
        else:
            if cnt > 0:
                s[i] = '1'
                possible, res = dfs(i - 1, cnt + 1)
                if possible != 1:
                    res = ''
            else:
                s[i] = '0'
                p0, res0 = dfs(i - 1, cnt)
                s[i] = '1'
                p1, res1 = dfs(i - 1, cnt + 1)
                possible = p0 + p1
                if p0 == 1 and p1 != 1:
                    res = res0
                elif p1 == 1 and p0 != 1:
                    res = res1
                else:
                    res = ''
        possible = 2 if possible >= 2 else possible
        return possible, res

    possible, ans = dfs(n, 0)
    ans = ans if possible == 1 else 'IMPOSSIBLE'
    answer(ans)
    dfs.cache_clear()

File: Codeforces/Div 3/Codeforces Round 988/2037E.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int query(int l, int r) {
    cout << "? " << l << " " << r << endl;
    cout.flush();
    int res;
    cin >> res;
    return res;
}

void answer(const string& ans) {
    cout << "! " << ans << endl;
    cout.flush();
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int t, n;
    cin >> t;
    while (t--) {
        cin >> n;

        vector<int> f(n + 2, 0), d(n + 1, 0);
        vector<char> s(n + 1);

        for (int i = 1; i < n; i++) {
            f[i] = query(i, n);
        }

        for (int i = 1; i <= n; i++) {
            d[i] = f[i] - f[i + 1];
        }

        bool ambiguous = false, found = false;
        string result;
        function<void(int, int)> dfs = [&](int i, int cnt) {
            if (found || ambiguous) return;
            if (i == 0) {
                found = true;
                result = string(s.begin() + 1, s.end());
                return;
            }
            if (d[i] < 0) return;
            if (d[i] > 0) {
                if (d[i] != cnt) return;
                s[i] = '0';
                dfs(i - 1, cnt);
            } else {
                int count = 0;
                s[i] = '0';
                dfs(i - 1, cnt);
                if (found) count++;
                found = false;

                s[i] = '1';
                dfs(i - 1, cnt + 1);
                if (found) count++;

                if (count == 0) return;
                if (count > 1) {
                    ambiguous = true;
                    return;
                }
            }
        };
        dfs(n, 0);

        if (!found || ambiguous) {
            answer("IMPOSSIBLE");
        } else {
            answer(result);
        }
    }
    return 0;
}
